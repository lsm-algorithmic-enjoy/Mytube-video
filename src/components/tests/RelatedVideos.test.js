import { Route } from "react-router-dom";
import { withAllContexts, withRouter } from "../../tests/utils";
import {
  screen,
  render,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { fakeVideos } from "../../tests/videos";
import RelatedVideos from "../RelatedVideos";

describe("RelatedVideos", () => {
  const fakeYoutube = {
    relatedVideos: jest.fn(),
  };
  afterEach(() => fakeYoutube.relatedVideos.mockReset());
  it("renders correctly", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
    const { asFragment } = renderRelatedVideos();

    await waitForElementToBeRemoved(screen.queryByText("Loading..."));
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders related videos correctly", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
    renderRelatedVideos();

    expect(fakeYoutube.relatedVideos).toHaveBeenCalledWith("id");
    await waitFor(() =>
      expect(screen.getAllByRole("listitem")).toHaveLength(fakeVideos.length)
    );
  });

  it("renders loading", () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
    renderRelatedVideos();

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("renders error", async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => {
      throw new Error("error");
    });
    renderRelatedVideos();

    await waitFor(() => {
      expect(screen.getByText("Something is wrong")).toBeInTheDocument();
    });
  });

  function renderRelatedVideos() {
    return render(
      withAllContexts(
        withRouter(<Route path="/" element={<RelatedVideos id="id" />} />),
        fakeYoutube
      )
    );
  }
});
