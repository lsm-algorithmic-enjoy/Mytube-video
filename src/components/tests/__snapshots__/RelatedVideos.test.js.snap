// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideos renders correctly 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="title"
        class="w-60 mr-2"
        src="http://image/"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title
        </p>
        <p
          class="text-sm opacity-80"
        >
          channelTitle
        </p>
        <p
          class="text-sm opacity-60"
        >
          just now
        </p>
      </div>
    </li>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="title2"
        class="w-60 mr-2"
        src="http://image2/"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2"
        >
          title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          channelTitle2
        </p>
        <p
          class="text-sm opacity-60"
        >
          just now
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
